sql/importer: TestImportIntoCSVCancel failed

sql/importer.TestImportIntoCSVCancel [failed](https://mesolite.cluster.engflow.com/invocations/default/998a0df8-2550-422f-9674-c8ddb76f5172?testReportRun=3&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvc3FsL2ltcG9ydGVyOmltcG9ydGVyX3Rlc3Q=) on master @ [2eeff2e7ccd8a50ea573e2c7c42a4829cde1e398](https://github.com/cockroachdb/cockroach/commits/2eeff2e7ccd8a50ea573e2c7c42a4829cde1e398):


```
=== RUN   TestImportIntoCSVCancel
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestImportIntoCSVCancel3557441559
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:181: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    import_stmt_test.go:1430: query 'SELECT count(*) FROM t': expected:
        0
        
        got:
        5000
        
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestImportIntoCSVCancel3557441559
--- FAIL: TestImportIntoCSVCancel (48.12s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestImportIntoCSVCancel.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58675

--- a/pkg/sql/importer/import_stmt_test.go
+++ b/pkg/sql/importer/import_stmt_test.go
@@ -1405,6 +1405,7 @@ func TestImportIntoCSVCancel(t *testing.T) {
 	conn := tc.ServerConn(0)
 
 	setupDoneCh := make(chan struct{})
+	finishedCancelCh := make(chan struct{})
 	for i := 0; i < tc.NumServers(); i++ {
 		tc.ApplicationLayer(i).JobRegistry().(*jobs.Registry).TestingWrapResumerConstructor(
 			jobspb.TypeImport,
@@ -1412,6 +1413,12 @@ func TestImportIntoCSVCancel(t *testing.T) {
 				r := raw.(*importResumer)
 				r.testingKnobs.onSetupFinish = func(flowinfra.Flow) {
 					close(setupDoneCh)
+				}
+				// We need this to ensure the cancel happens before the descriptor is
+				// marked online, and thus the rows ingested will be reverted.
+				r.testingKnobs.afterImport = func(_ roachpb.RowCount) error {
+					<-finishedCancelCh
+					return nil
 				}
 				return r
 			})
@@ -1427,6 +1434,7 @@ func TestImportIntoCSVCancel(t *testing.T) {
 	row.Scan(&jobID)
 	<-setupDoneCh
 	sqlDB.Exec(t, fmt.Sprintf("CANCEL JOB %d", jobID))
+	close(finishedCancelCh)
 	sqlDB.Exec(t, fmt.Sprintf("SHOW JOB WHEN COMPLETE %d", jobID))
 	sqlDB.CheckQueryResults(t, "SELECT count(*) FROM t", [][]string{{"0"}})
 

